blsftcosi/protocol/sub_protocol.go

```diff
@@ -0,0 +1,431 @@
+	defer func() {
+		err := p.Broadcast(&Stop{})
+		if err != nil {
+			log.Error("error while broadcasting stopping message:", err)
+		}
+	}()
+
+	// make sure we're ready to go
+	hasStarted := <-p.startChan
+	if !hasStarted {
+		return nil
+	}
+
+	// Only one child anyway
+	err := p.SendToChildren(&Announcement{
+		Msg:       p.Msg,
+		Data:      p.Data,
+		Timeout:   p.Timeout,
+		Threshold: p.Threshold,
+	})
+	if err != nil {
+		// Only log what happened so we can try to finish the protocol
+		// (e.g. one child is offline)
+		log.Warnf("Error when broadcasting to children: %s", err.Error())
+	}
+
+	select {
+	case <-p.closeChan:
+		return nil
+	case reply := <-p.ChannelResponse:
+		if reply.Equal(p.Root().Children[0]) {
+			// Transfer the response to the parent protocol
+			p.subResponse <- reply
+		}
+	case <-time.After(p.Timeout):
+		// It might be only the subleader then we send a notification
+		// to let the parent protocol take actions
+		log.Warn(p.ServerIdentity(), "timed out while waiting for subleader response")
+		p.subleaderNotResponding <- true
+	}
+
+	return nil
+}
+
+// dispatchSubLeader takes care of synchronizing the children
+// responses and aggregate them to eventually send that to
+// the root
+func (p *SubBlsFtCosi) dispatchSubLeader() error {
+	a := p.waitAnnouncement(p.Root())
+	if a == nil {
+		return nil
+	}
+
+	errs := p.SendToChildrenInParallel(a)
+	if len(errs) > 0 {
+		log.Error(errs)
+	}
+
+	responses := make(ResponseMap)
+	for _, c := range p.Children() {
+		// Accept response for those identities only
+		responses[c.ID] = nil
+	}
+
+	own, err := p.makeResponse()
+	if ok := p.verificationFn(p.Msg, p.Data); ok {
+		log.Lvlf3("Subleader %v signed", p.ServerIdentity())
+		responses[p.TreeNode().ID] = own
+	}
+
+	// we need to timeout the children faster than the root timeout to let it
+	// know the subleader is alive, but some children are failing
+	timeout := time.After(p.Timeout / 2)
+	done := 0
+	for done < len(p.Children()) {
+		select {
+		case <-p.closeChan:
+			return nil
+		case reply, ok := <-p.ChannelResponse:
+			r, ok := responses[reply.ID]
+			if !ok {
+				log.Warnf("Got a message from an unknown node %v", reply.ServerIdentity.ID)
+			} else if r == nil {
+				public := p.NodePublic(reply.ServerIdentity)
+				if err := bls.Verify(p.suite, public, p.Msg, reply.Signature); err == nil {
+					responses[reply.ID] = &reply.Response
+					done++
+				}
+			} else {
+				log.Warnf("Duplicate message from %v", reply.ServerIdentity.ID)
+			}
+		case reply := <-p.ChannelRefusal:
+			_, ok := responses[reply.ID]
+			serviceName := onet.ServiceFactory.Name(p.Token().ServiceID)
+
+			if !ok {
+				log.Warnf("Got a message from an unknown node %v", reply.ServerIdentity.ID)
+			} else if err := bls.Verify(p.suite, reply.ServerIdentity.ServicePublic(serviceName), reply.Nonce, reply.Signature); err == nil {
+				// The child gives an empty signature as a mark of refusal
+				responses[reply.ID] = &Response{}
```

Comment:
I have a feeling that using the treeID as the key might be a security risk. Is the treeID verified by onet before it gets to this point? In order words, can malicious nodes forget the treeID? It might be safer to use the public key as the key to `responses` because at you can verify that they actually hold the private key by checking the signature.